src/compiler/checker.ts

```diff
@@ -5696,6 +5696,29 @@ module ts {
             if (!links.resolvedType) {
                 links.resolvedType = checkExpression(node.expression);
 
+                // Disallow using a static property in computedPropertyName because classDeclaration is bound lexically in ES6
+                // and its static property assignment will be emitted after classDeclaration.
+                // Therefore, using static property inside computedPropertyName will cause an use-before-definition error
+                // Example:
+                //  * TypeScript
+                //      class C {
+                //          static p = 10;
+                //          [C.p]() {}
+                //      }
+                //  * JavaScript
+                //      class C {
+                //          [C.p]() {}  // Use before definition error
+                //      }
+                //      C.p = 10;
+                if (links.resolvedSymbol) {
+                    var declarations = links.resolvedSymbol.declarations;
+                    forEach(declarations, declaration => {
+                        if (declaration.flags & NodeFlags.Static) {
+                            error(node, Diagnostics.A_computed_property_name_cannot_reference_a_static_property);

```

Comment:
Also, this is not general, because it doesn't handle something like:
```ts
class C {
    static hello = "hello";
    static goodbye = "goodbye";
    [hello + goodbye]() { }
}
```